diff --git a/configs/Base-RCNN-FPN.yaml b/configs/Base-RCNN-FPN.yaml
index 3e020f2..91b16c6 100644
--- a/configs/Base-RCNN-FPN.yaml
+++ b/configs/Base-RCNN-FPN.yaml
@@ -17,7 +17,7 @@ MODEL:
     # (See "modeling/rpn/rpn_outputs.py" for details of this legacy issue)
     # which is approximately 1000 proposals per-image since the default batch size for FPN is 2.
     POST_NMS_TOPK_TRAIN: 1000
-    POST_NMS_TOPK_TEST: 1000
+    POST_NMS_TOPK_TEST: 200
   ROI_HEADS:
     NAME: "StandardROIHeads"
     IN_FEATURES: ["p2", "p3", "p4", "p5"]
diff --git a/detectron2/data/transforms/augmentation_impl.py b/detectron2/data/transforms/augmentation_impl.py
index db727cd..bfef4a4 100644
--- a/detectron2/data/transforms/augmentation_impl.py
+++ b/detectron2/data/transforms/augmentation_impl.py
@@ -173,8 +173,10 @@ class ResizeShortestEdge(Augmentation):
             scale = self.max_size * 1.0 / max(newh, neww)
             newh = newh * scale
             neww = neww * scale
-        neww = int(neww + 0.5)
-        newh = int(newh + 0.5)
+        # neww = int(neww + 0.5)
+        # newh = int(newh + 0.5)
+        newh = 416
+        neww = 416
         return ResizeTransform(h, w, newh, neww, self.interp)
 
 
