--- ./yolov5/utils/downloads.py	2021-10-29 12:12:25.527251946 +0530
+++ downloads-modified.py	2021-10-29 12:09:35.436000000 +0530
@@ -67,6 +67,13 @@
             except:
                 tag = 'v5.0'  # current release
 
+        print("\n\n")
+        print("=="*30)
+        print("Changing the release tag from latest to v5.0 to download older checkpoint.")
+        tag = 'v5.0'
+        print("=="*30)
+        print("\n\n")
+
         if name in assets:
             safe_download(file,
                           url=f'https://github.com/{repo}/releases/download/{tag}/{name}',
