--- ./yolov5/models/yolo.py	2021-10-26 15:17:56.907487210 +0530
+++ yolo.py	2021-10-26 15:04:18.566018710 +0530
@@ -32,6 +32,21 @@
 LOGGER = logging.getLogger(__name__)
 
 
+def filter_boxes(box_xywh):
+    box_xy, box_wh = torch.split(box_xywh, [2, 2], dim=-1)
+
+    box_mins = (box_xy - (box_wh / 2.))
+    box_maxes = (box_xy + (box_wh / 2.))
+    boxes = torch.cat((
+        box_mins[..., 1:2],  # y_min
+        box_mins[..., 0:1],  # x_min
+        box_maxes[..., 1:2],  # y_max
+        box_maxes[..., 0:1]  # x_max
+    ), dim=-1)
+
+    return boxes
+
+
 class Detect(nn.Module):
     stride = None  # strides computed during build
     onnx_dynamic = False  # ONNX export parameter
@@ -49,13 +64,20 @@
         self.m = nn.ModuleList(nn.Conv2d(x, self.no * self.na, 1) for x in ch)  # output conv
         self.inplace = inplace  # use in-place ops (e.g. slice assignment)
 
-    def forward(self, x):
+    def forward(self, x, without_abp):
         z = []  # inference output
+        list_boxes = []
+        list_scores = []
+
         for i in range(self.nl):
             x[i] = self.m[i](x[i])  # conv
             bs, _, ny, nx = x[i].shape  # x(bs,255,20,20) to x(bs,3,20,20,85)
-            x[i] = x[i].view(bs, self.na, self.no, ny, nx).permute(0, 1, 3, 4, 2).contiguous()
+            # x[i] = x[i].view(bs, self.na, self.no, ny, nx).permute(0, 1, 3, 4, 2).contiguous()
+            x[i] = x[i].view(-1, self.na, self.no, ny, nx).permute(0, 1, 3, 4, 2).contiguous()
 
+            if without_abp:
+                continue
+                
             if not self.training:  # inference
                 if self.grid[i].shape[2:4] != x[i].shape[2:4] or self.onnx_dynamic:
                     self.grid[i] = self._make_grid(nx, ny).to(x[i].device)
@@ -64,14 +86,47 @@
                 if self.inplace:
                     y[..., 0:2] = (y[..., 0:2] * 2. - 0.5 + self.grid[i]) * self.stride[i]  # xy
                     y[..., 2:4] = (y[..., 2:4] * 2) ** 2 * self.anchor_grid[i]  # wh
+                    # z.append(y.view(bs, -1, self.no))
+                    z.append(y.view(-1, self.na * ny * nx, self.no))
                 else:  # for YOLOv5 on AWS Inferentia https://github.com/ultralytics/yolov5/pull/2953
                     xy = (y[..., 0:2] * 2. - 0.5 + self.grid[i]) * self.stride[i]  # xy
                     wh = (y[..., 2:4] * 2) ** 2 * self.anchor_grid[i].view(1, self.na, 1, 1, 2)  # wh
-                    y = torch.cat((xy, wh, y[..., 4:]), -1)
-                z.append(y.view(bs, -1, self.no))
-
-        return x if self.training else (torch.cat(z, 1), x)
-
+                    xywh = torch.cat((xy,wh),4)
+                    scores = y[:,:,:,:,4:]
+                    # scores : [B x 3 x h x w x 81]
+
+                    # list_boxes.append(xywh.view(bs, -1, 4))
+                    list_boxes.append(xywh.view(-1, self.na * ny * nx, 4))
+                    # list_scores.append(scores.view(bs, -1, 81))
+                    list_scores.append(scores.view(-1, self.na * ny * nx, 81))
+
+
+        if self.inplace:
+            return x if self.training else (torch.cat(z, 1), x)
+        elif without_abp:
+            return x
+        else:
+            all_boxes = torch.cat(list_boxes, 1)         # [B x N x 4]
+            all_scores = torch.cat(list_scores, 1)       # [B x N x 81]
+
+            obj_prob = all_scores[:,:,0:1]
+            # obj_prob : [B x N x 1]
+
+            cls_prob = all_scores[:,:,1:]
+            # cls_prob : [B x N x 80]
+
+            all_scores_final = obj_prob * cls_prob
+            # scores : [B x N x 80]
+
+            all_boxes = filter_boxes(all_boxes)     
+            # convert from xyxy to yxyx
+            # As TF NMS operator needs bbox in yxyx.
+            # all_boxes = all_boxes.view(bs, -1, 1, 4)
+            all_boxes = torch.unsqueeze(all_boxes, 2)
+
+            return all_boxes, all_scores_final
+            
+            
     @staticmethod
     def _make_grid(nx=20, ny=20):
         yv, xv = torch.meshgrid([torch.arange(ny), torch.arange(nx)])
@@ -116,6 +171,35 @@
         initialize_weights(self)
         self.info()
         LOGGER.info('')
+        
+        # Setting default values for nms params
+        # This will be overridden by set_nms_params
+        self.apply_nms=False
+        self.use_pytorch_custom_op = False
+        self.max_output_boxes_per_class = 26
+        self.max_total_size_quantity = 2080
+        self.iou_thresh=0.65
+        self.score_thresh=0.001
+        self.clip_boxes = False
+        self.pad_per_class = False
+        self.max_det=300 
+
+    # This method should be called before calling the forward method of this class.
+    # This would override the export related parameters and use it while constructing 
+    # nms operations which in turn gets hard coded in the model.
+    def set_export_params(self, apply_nms=False, use_pytorch_custom_op=False, \
+                max_output_boxes_per_class=26, \
+                iou_thresh=0.65, score_thresh=0.001, \
+                clip_boxes=False, pad_per_class=False, max_det=2080):
+        self.apply_nms = apply_nms
+        self.use_pytorch_custom_op = use_pytorch_custom_op
+        self.max_output_boxes_per_class = max_output_boxes_per_class
+        self.max_total_size_quantity = max_det
+        self.iou_thresh = iou_thresh
+        self.score_thresh = score_thresh
+        self.clip_boxes = clip_boxes
+        self.pad_per_class = pad_per_class
+
 
     def forward(self, x, augment=False, profile=False, visualize=False):
         if augment:
@@ -142,11 +226,66 @@
                 x = y[m.f] if isinstance(m.f, int) else [x if j == -1 else y[j] for j in m.f]  # from earlier layers
             if profile:
                 self._profile_one_layer(m, x, dt)
-            x = m(x)  # run
+            if isinstance(m, Detect):
+                without_abp = not self.apply_nms
+                x = m(x, without_abp)
+            else:
+                x = m(x)
             y.append(x if m.i in self.save else None)  # save output
             if visualize:
                 feature_visualization(x, m.type, m.i, save_dir=visualize)
-        return x
+        
+        if not self.inplace:
+            # if self.include_abp_only:
+            #     all_boxes, all_scores = x[0], x[1]
+            #     # all_boxes : [B x N x 1 x 4]
+            #     # all_scores : [B x N x 80]
+                
+            #     # This would create model with ABP part only and the nodes which 
+            #     # are related to QNMS are stiched out side of this class.
+            #     return all_boxes, all_scores
+            if self.apply_nms:
+                # With ABP QNMS variant
+                all_boxes, all_scores = x[0], x[1]
+                # all_boxes : [B x N x 1 x 4]
+                # all_scores : [B x N x 80]
+
+                if self.use_pytorch_custom_op:
+                    # This would add QNMS node in the graph. 
+                    # Both the ONNX and Pytorch model would contain the QNMS node.
+                    import sys
+                    sys.path.append("../../")
+                    from yolo_qnms_plugin import CustomQnmsYolo
+                    NMSmodule = CustomQnmsYolo(self.clip_boxes, self.pad_per_class)
+
+                    res_boxes, res_scores, res_classes, res_num_boxes = \
+                                        NMSmodule(all_boxes, all_scores, \
+                                                    self.max_output_boxes_per_class, \
+                                                    self.max_total_size_quantity, \
+                                                    self.iou_thresh, self.score_thresh)
+
+                    # res_boxes : [1 x 100 x 4] : float32
+                    # res_scores : [1 x 100] : float32
+                    # res_classes : [1 x 100] : int32
+                    # res_num_boxes : [1] : int32
+
+                    return res_boxes, res_scores, res_classes, res_num_boxes
+                else:
+                    # This would return only ABP output and 
+                    # stitch QNMS node in exported ONNX graph.
+                    # Only ONNX model is prepared here.
+                    return all_boxes, all_scores
+            else:
+                # Without ABP NMS variant
+                # feature_map_1 : [1, 3, 80, 80, 85]
+                # feature_map_2 : [1, 3, 40, 40, 85]
+                # feature_map_3 : [1, 3, 20, 20, 85]
+
+                return x
+        else:
+            # Return from original workflow of yolo
+            return x
+
 
     def _descale_pred(self, p, flips, scale, img_size):
         # de-scale predictions following augmented inference (inverse operation)
